dist/background.js

```diff
@@ -2,63 +2,129 @@
+    this.prefs = prefs // shorthand language strings ['au', 'de']
+    this.langs = languages // language strings ['en-au', 'de-de']
+    this.ownWords = ownWords // word strings ['kablam', 'shebang']
+    this.spells = this.createSpellCheckers() // nspell instances
   }
 
-  const spellCheckers = {}
+  // create spell checkers in order of languages specified by user
+  createSpellCheckers () {
+    if (this.prefs.length !== this.dicts.length) {
+      console.warn('MultiDict: Language prefs and user dictionary length not equal. Aborting.')
+      return
+    }
 
-  for (let i = 0; i < user.dicts.length; i++) {
-    spellCheckers[languagePrefs[i]] = nspell(user.dicts[i])
+    const spells = {}
+    for (let i = 0; i < this.dicts.length; i++) {
+      spells[this.prefs[i]] = nspell(this.dicts[i])
+    }
+    return spells
   }
 
-  return spellCheckers
+  // add word user.ownWords and add also to existing spell checker instances
+  addWord (word) {
+    user.ownWords.push(word)
+    Object.keys(this.spells).forEach((language) => {
+      this.spells[language].add(word)
+    })
+  }
 }
 
-// handles all incoming messages from the content script
-function listener (message) {
-  messageHandler = message
-  messageHandler.postMessage({ greeting: 'MultDict connection established' })
-
-  // content script message object should contain: { name, language, content }
-  messageHandler.onMessage.addListener((message) => {
-    if (message.language === 'unreliable') {
+// expected: { type: '', detectedLanguage: '', content: '' }
+// checkSpelling returns: { cleanedContent: [], misspeltWords: [], suggestions: {}}
+function check (message) {
+  switch (message.detectedLanguage) {
+    case 'unreliable':
       messageHandler.postMessage({
-        suggestions: checkSpelling(spells[languagePrefs[0]], message.content)
+        type: 'spelling',
+        suggestions: checkSpelling(user.spells[user.prefs[0]], message.content)
       })
-    } else {
-      for (const pref of languagePrefs) {
-        if (languages.includes(`${message.language}-${pref}`)) {
+      break
+
+    default:
+      for (const pref of user.prefs) {
+        if (user.langs.includes(`${message.detectedLanguage}-${pref}`)) {
           messageHandler.postMessage({
-            spelling: checkSpelling(spells[pref], message.content)
+            type: 'spelling',
+            spelling: checkSpelling(user.spells[pref], message.content)
           })
           break
         }
       }
-    }
-  })
+  }
+}
+
+// stores a word in browser sync storage (automatically falls back to local) and calls user.addWord
+async function storeCustomWord (word) {
+  console.log('storing word: ' + word)
+  if (customWords.includes(word)) {
+    console.warn(`MultiDict: custom dictionary contains has word : "${word}". Aborting.`)
+    return
+  }
+
+  customWords.push(word)
+  const error = await browser.storage.sync.set({ personal: customWords })
```

Comment:
Maybe worth finding a way to append the customWord to the stored array of custom words instead of overwriting the whole thing each time...